File: javascript/aoc_toolbox.js
const fs = require('fs')

class InputFile {
    constructor(filename = "input.txt") {
        this._lines = fs.readFileSync(filename).toString().trim().split("\n")
    }

    lines() {
        return this._lines
    }

    linesInts() {
        return this.lines().map((x) => parseInt(x))
    }

    linefieldsRegexp(r) {
        return this.lines().map((line) => {
            let matches = line.match(r)
            matches.shift()
            return matches
        })
    }

    sections() {
        let sections = []
        let currentSection = []
        let ls = this.lines()
        for (let lineNum in ls) {
            if (ls[lineNum] == '') {
                if (currentSection.length > 0) {
                    sections.push(currentSection)
                }
                currentSection = []
            } else {
                currentSection.push(ls[lineNum])
            }
        }
        if (currentSection.length > 0) {
            sections.push(currentSection)
        }
        return sections
    }

    matrixChar(row, col) {
        return this.lines()[row][col]
    }

    subMatrix(startRow, startCol, rows, cols) {
        let sub = []
        let ls = this.lines()
        for (let rowNum = startRow; rowNum < startRow + rows; rowNum++) {
            let subRow = ls[rowNum].substr(startCol, cols)
            sub.push(subRow)
        }
        return sub
    }
}

module.exports = InputFile

File: ruby/aoc_toolbox.rb
class InputData
    def initialize(filename: 'input.txt', lines: nil, strip_newline: true)
        if (lines)
            @raw_lines = lines.dup
        else
            @filename = filename
            @raw_lines = File.open(@filename).readlines()
            @raw_lines.map!{|line| line.strip} if strip_newline
        end
    end

    public
    # returns an array of lines as strings. each line is stripped from ending newline char.
    def lines
        @raw_lines
    end

    # returns an array of lines as integers
    def lines_ints
        self.lines.map(&:to_i)
    end

    # returns an array of lines, each containing an array of 
    # fields identified by the provided regexp
    def linefields_regexp(rx)
        self.lines.map{|line| line.match(rx).to_a.drop(1)}
    end

    # returns an array of sections, as identified by the existance of 
    # empty lines in the input data. Each section is a InputData instance containing 
    # the lines for each section.
    def sections
        sections = []
        current_section = []
        self.lines.each do |line|
            if line == ''
                sections.append(InputData.new(lines: current_section)) unless current_section.empty?
                current_section = []
            else
                current_section.append(line)
            end
        end
        # deal with last line if non-empty
        sections.append(InputData.new(lines: current_section)) unless current_section.empty?
        sections
    end

    def matrix_char(row, col)
        self.lines[row][col]
    end

    # returns a InputData instance representing the sub matrix pointed to by the parameters
    def sub_matrix(start_row, start_col, rows, cols)
        sub = []
        for row in self.lines[start_row...(start_row+rows)]
            sub_row = row[start_col...(start_col+cols)]
            sub.append(sub_row)
        end
        InputData.new(lines: sub)
    end
end
